cpp/src/slice2cs/CsUtil.cpp

```diff
@@ -1103,223 +1105,123 @@ Slice::CsGenerator::sequenceUnmarshalCode(const SequencePtr& seq, const string&
-            }
-            else if(st->minWireSize() > 1)
-            {
-                out << nl << stream << ".SkipSize();";
-            }
-            string tmp = "tmpVal";
-            out << nl << seqS << ' ' << tmp << ';';
-            writeUnmarshalCode(out, seq, scope, tmp, stream);
-            if(isArray)
-            {
-                out << nl << param << " = " << tmp << ";";
-            }
-            else
-            {
-                out << nl << param << " = new " << seqS << "(" << tmp << ");";
-            }
-            out << eb;
-            out << nl << "else";
-            out << sb;
-            out << nl << param << " = null;";
-            out << eb;
-        }
-        return;
+        return false;
     }
 
-    //
-    // At this point, all remaining element types have variable size.
-    //
-    if(marshal)
+    return true;
+}
+
+}
+void
+Slice::CsGenerator::writeTaggedSequenceMarshalCode(Output& out,
+                                                   const SequencePtr& seq,
+                                                   const string& scope,
+                                                   const string& param,
+                                                   int tag,
+                                                   const string& stream)
+{
+    const TypePtr type = seq->type();
+    const bool isArray = !seq->hasMetaDataWithPrefix("cs:generic:");
+    const string length = isArray ? param + ".Length" : param + ".Count";
+    BuiltinPtr builtin = BuiltinPtr::dynamicCast(type);
+    StructPtr st = StructPtr::dynamicCast(type);
+
+    if(isVariableLength(type))
     {
-        out << nl << "if(" << param << " != null && " << stream << ".WriteOptional(" << tag << ", "
+        out << nl << "if (" << param << " != null && " << stream << ".WriteOptional(" << tag << ", "
             << getTagFormat(seq, scope) << "))";
         out << sb;
         out << nl << "var pos = " << stream << ".StartSize();";
         writeMarshalCode(out, seq, scope, param, stream);
         out << nl << stream << ".EndSize(pos);";
         out << eb;
     }
-    else
+    else if(st)
     {
-        out << nl << "if(" << stream << ".ReadOptional(" << tag << ", " << getTagFormat(seq, scope) << "))";
+        out << nl << "if (" << param << " != null && " << stream << ".WriteOptional(" << tag << ", "
+            << getTagFormat(seq, scope) << "))";
         out << sb;
-        out << nl << stream << ".Skip(4);";
-        string tmp = "tmpVal";
-        out << nl << seqS << ' ' << tmp << ';';
-        writeUnmarshalCode(out, seq, scope, tmp, stream);
-        if(isArray)
+        if(st->minWireSize() > 1)
         {
-            out << nl << param << " = " << tmp << ";";
-        }
-        else
-        {
-            out << nl << param << " = new " << seqS << "(" << tmp << ");";
+            out << nl << stream << ".WriteSize(" << length << " * " << st->minWireSize() << " + (" << length
```

Comment:
Should add a comment. This works because the type is not variable length and minWireSize is the fixed-size of each element.

Reply:
Added a comment